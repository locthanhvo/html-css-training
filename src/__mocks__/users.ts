import { STATUS } from '@/constants';

export const USERS = Array.from({ length: 5 }).map((_, index) => ({
  id: `${index}`,
  firstName: 'Devon',
  lastName: 'Lane',
  email: 'devon@mail.com',
  phone: '+237 6 99 88 77 66',
  status: STATUS.VALID,
  createdAt: '2024-06-05 14:46:58',
  updatedAt: '2024-06-05 14:46:58',
}));
