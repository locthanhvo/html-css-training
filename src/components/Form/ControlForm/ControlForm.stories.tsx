import { StoryObj, Meta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import ControlForm from '.';

const meta: Meta<typeof ControlForm> = {
  title: 'Components/ControlForm',
  tags: ['autodocs'],
  component: ControlForm,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof ControlForm>;

export const AddControl: Story = {
  args: {},
};

export const EditControl: Story = {
  args: {
    title: 'Edit Control Teams',
    isEdit: true,
  },
};
