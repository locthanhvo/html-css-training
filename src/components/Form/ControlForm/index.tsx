import { memo, useCallback } from 'react';
import { Flex, Text } from '@chakra-ui/react';
import { Controller, useForm } from 'react-hook-form';

// Components
import { InputField, ActionButton, TextareaField } from '@/components';

// Types
import { TControl } from '@/types';

// Utils
import { formatDateString } from '@/utils';

// Constants
import { CONTROL_SCHEMA } from '@/constants';

interface ControlFormProps {
  title?: string;
  isEdit?: boolean;
  isLoading?: boolean;
  initialValues?: TControl;
  onSubmit: (data: TControl) => void;
  onCancel?: () => void;
}

const ControlForm = ({
  title = 'Add Control Teams',
  initialValues,
  isLoading = false,
  isEdit = false,
  onSubmit,
  onCancel,
}: ControlFormProps) => {
  const {
    control,
    formState: { isSubmitting, isValid, isDirty },
    handleSubmit,
    clearErrors,
  } = useForm<TControl>({
    mode: 'onChange',
    reValidateMode: 'onChange',
    defaultValues: {
      name: initialValues?.name ?? '',
      description: initialValues?.description ?? '',
      createdAt:
        formatDateString(initialValues?.createdAt) ||
        formatDateString(new Date().toISOString()),
      updatedAt:
        formatDateString(initialValues?.updatedAt) ||
        formatDateString(new Date().toISOString()),
    },
  });

  const handleInputChange = useCallback(
    ({
      field,
      data,
      isError,
      onChange,
    }: {
      field: keyof TControl;
      data: string;
      isError: boolean;
      onChange: (value: string) => void;
    }) => {
      isError && clearErrors(field);
      onChange(data);
    },
    [clearErrors],
  );

  return (
    <Flex
      as="form"
      w="full"
      pb={5}
      direction="column"
      gap={4}
      onSubmit={handleSubmit(onSubmit)}
    >
      <Text fontWeight="bold" fontSize="md">
        {title}
      </Text>

      <Flex direction="column" gap={3}>
        <Controller
          key="name"
          control={control}
          rules={CONTROL_SCHEMA.NAME}
          name="name"
          render={({ field, fieldState: { error } }) => (
            <InputField
              label="Name"
              placeholder="Please enter name"
              {...field}
              isError={!!error}
              errorMessages={error?.message}
              isDisabled={isSubmitting}
              onChange={(data) =>
                handleInputChange({
                  field: 'name',
                  data,
                  isError: !!error,
                  onChange: field.onChange,
                })
              }
              aria-label="name"
            />
          )}
        />

        <Controller
          key="description"
          control={control}
          rules={CONTROL_SCHEMA.DESCRIPTION}
          name="description"
          render={({ field, fieldState: { error } }) => (
            <TextareaField
              label="Description"
              placeholder="Please enter description..."
              {...field}
              isError={!!error}
              errorMessages={error?.message}
              isDisabled={isSubmitting}
              onChange={(data) =>
                handleInputChange({
                  field: 'description',
                  data,
                  isError: !!error,
                  onChange: field.onChange,
                })
              }
              aria-label="name"
            />
          )}
        />
      </Flex>

      <Flex gap={3} mt={10}>
        <ActionButton
          type="submit"
          size="lg"
          title={isEdit ? 'Update' : 'Create'}
          bgColor="blue.200"
          isDisabled={!isDirty || (!isValid && !isSubmitting)}
          isLoading={isLoading || isSubmitting}
        />

        <ActionButton
          type="button"
          size="lg"
          title="Cancel"
          bgColor="primary"
          color="black"
          onClick={onCancel}
        />
      </Flex>
    </Flex>
  );
};

export default memo(ControlForm);
