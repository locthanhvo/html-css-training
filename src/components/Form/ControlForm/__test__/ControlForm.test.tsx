import { render, screen } from '@testing-library/react';

// Component
import ControlForm from '..';
import { BrowserRouter } from 'react-router-dom';
import { USERS } from '@/__mocks__';

const mockOnSubmit = jest.fn();

describe('ControlForm components', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('match snapshot with add form', () => {
    const { container } = render(
      <BrowserRouter>
        <ControlForm onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('match snapshot with edit form', () => {
    const { container } = render(
      <BrowserRouter>
        <ControlForm
          title="Edit Control"
          isEdit
          initialValues={USERS[0]}
          onSubmit={mockOnSubmit}
        />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders Add form by default', () => {
    render(
      <BrowserRouter>
        <ControlForm onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(screen.getByRole('button', { name: 'Create' })).toBeInTheDocument();
  });

  it('renders Edit form by default', () => {
    render(
      <BrowserRouter>
        <ControlForm title="Edit Control" isEdit onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(screen.getByRole('button', { name: 'Update' })).toBeInTheDocument();
  });
});
