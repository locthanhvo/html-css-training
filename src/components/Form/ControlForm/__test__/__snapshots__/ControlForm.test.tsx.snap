// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlForm components match snapshot with add form 1`] = `
<div>
  <form
    class="css-1urdh7u"
  >
    <p
      class="chakra-text css-5fbnf6"
    >
      Add Control Teams
    </p>
    <div
      class="css-1pyx7hi"
    >
      <div
        aria-label="name"
        class="chakra-form-control css-0"
        name="name"
        placeholder="Please enter name"
        role="group"
        value=""
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Name
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="name"
            class="chakra-input css-adz8kx"
            id="field-:r0:"
            name="name"
            placeholder="Please enter name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Description
        </label>
        <textarea
          aria-label="name"
          class="chakra-textarea css-mie5s1"
          id="field-:r1:"
          name="description"
          placeholder="Please enter description..."
        />
      </div>
    </div>
    <div
      class="css-bbgbcw"
    >
      <button
        class="chakra-button css-ey593w"
        disabled=""
        name="Create"
        type="submit"
      >
        Create
      </button>
      <button
        class="chakra-button css-15qijou"
        name="Cancel"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;

exports[`ControlForm components match snapshot with edit form 1`] = `
<div>
  <form
    class="css-1urdh7u"
  >
    <p
      class="chakra-text css-5fbnf6"
    >
      Edit Control
    </p>
    <div
      class="css-1pyx7hi"
    >
      <div
        aria-label="name"
        class="chakra-form-control css-0"
        name="name"
        placeholder="Please enter name"
        role="group"
        value=""
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          Name
        </label>
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            aria-label="name"
            class="chakra-input css-adz8kx"
            id="field-:r2:"
            name="name"
            placeholder="Please enter name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-zecyy3"
          for="field-:r3:"
          id="field-:r3:-label"
        >
          Description
        </label>
        <textarea
          aria-label="name"
          class="chakra-textarea css-mie5s1"
          id="field-:r3:"
          name="description"
          placeholder="Please enter description..."
        />
      </div>
    </div>
    <div
      class="css-bbgbcw"
    >
      <button
        class="chakra-button css-ey593w"
        disabled=""
        name="Update"
        type="submit"
      >
        Update
      </button>
      <button
        class="chakra-button css-15qijou"
        name="Cancel"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
